File: backend/app/constants/mood.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Dict, Optional

PROJECT_ROOT = Path(__file__).resolve().parents[3]
CONFIG_PATH = PROJECT_ROOT / "frontend" / "config" / "mood.json"

_DEFAULT_MAP: Dict[int, str] = {
    -2: "最悪",
    -1: "やや悪",
    0: "普通",
    1: "やや良",
    2: "最高",
}
_DEFAULT_LABEL = "不明"

def _load_config() -> Dict[int, str]:
    try:
        with CONFIG_PATH.open("r", encoding="utf-8") as f:
            data = json.load(f)
        moods = data.get("moods", [])
        mapping = {int(item["value"]): item.get("short_label") or item.get("label") for item in moods if "value" in item}
        return {k: v for k, v in mapping.items() if v}
    except Exception:
        return {}

_CONFIG_MAP = _load_config()
_LABEL_MAP: Dict[int, str] = {**_DEFAULT_MAP, **_CONFIG_MAP}
_FALLBACK_LABEL = _DEFAULT_LABEL

def get_mood_label(score: Optional[int]) -> str:
    try:
        key = int(score) if score is not None else None
    except Exception:
        key = None
    return _LABEL_MAP.get(key, _FALLBACK_LABEL)

File: frontend/lib/mood.ts
import type { MoodOption } from "./types";

import moodConfig from "../config/mood.json";

type MoodConfigItem = { value: number; label: string; short_label?: string };

const parsed = (moodConfig as { moods?: MoodConfigItem[] }).moods ?? [];

export const moodOptions: MoodOption[] =
  parsed.length > 0
    ? parsed.map((m) => ({
        value: m.value,
        label: m.label,
      }))
    : [
        { value: -2, label: "とてもネガティブ (-2)" },
        { value: -1, label: "ややネガティブ (-1)" },
        { value: 0, label: "ニュートラル (0)" },
        { value: 1, label: "ややポジティブ (+1)" },
        { value: 2, label: "とてもポジティブ (+2)" },
      ];

const labelMap = new Map<number, { label: string; shortLabel: string }>();
const DEFAULT_LABEL = (moodConfig as { default_label?: string }).default_label ?? "不明";
const DEFAULT_SHORT_LABEL = (moodConfig as { default_short_label?: string }).default_short_label ?? "不明";

parsed.forEach((m) => {
  labelMap.set(m.value, {
    label: m.label,
    shortLabel: m.short_label ?? m.label,
  });
});

const toNumber = (score: number | null | undefined) => {
  const n = Number(score);
  return Number.isFinite(n) ? n : null;
};

export const getMoodLabel = (score: number | null | undefined): string => {
  const key = toNumber(score);
  if (key !== null && labelMap.has(key)) {
    return labelMap.get(key)!.label;
  }
  return DEFAULT_LABEL;
};

export const getMoodShortLabel = (score: number | null | undefined): string => {
  const key = toNumber(score);
  if (key !== null && labelMap.has(key)) {
    return labelMap.get(key)!.shortLabel;
  }
  return DEFAULT_SHORT_LABEL;
};

